wpi.sh

```diff
@@ -0,0 +1,177 @@
+fi
+
+if [ ! -d "${JAVA8_HOME}" ]; then
+    echo "JAVA8_HOME is set to a non-existent directory ${JAVA8_HOME}"
+    exit 1
+fi
+
+# Testing for JAVA11_HOME, not a misspelling of JAVA_HOME.
+# shellcheck disable=SC2153
+if [ "x${JAVA11_HOME}" = "x" ]; then
+    echo "JAVA11_HOME must be set to a Java 11 JDK"
+    exit 1
+fi
+
+if [ ! -d "${JAVA11_HOME}" ]; then
+    echo "JAVA11_HOME is set to a non-existent directory ${JAVA11_HOME}"
+    exit 1
+fi
+
+JAVA_HOME="${JAVA11_HOME}"
+
+if [ "x${CHECKERFRAMEWORK}" = "x" ]; then
+    echo "CHECKERFRAMEWORK is not set; it must be set to a locally-built Checker Framework. Please clone and build github.com/typetools/checker-framework"
+    exit 2
+fi
+
+if [ ! -d "${CHECKERFRAMEWORK}" ]; then
+    echo "CHECKERFRAMEWORK is set to a non-existent directory ${CHECKERFRAMEWORK}"
+    exit 2
+fi
+
+if [ "x${DIR}" = "x" ]; then
+    echo "wpi.sh was called without a -d argument. The -d argument must be the absolute path to the directory containing the project on which to run WPI."
+    exit 4
+fi
+
+if [ ! -d "${DIR}" ]; then
+    echo "wpi.sh's -d argument was not a directory: ${DIR}"
+    exit 4
+fi
+
+function configure_and_exec_dljc {
+
+  if [ -f build.gradle ]; then
+      if [ -f gradlew ]; then
+	  chmod +x gradlew
+	  GRADLE_EXEC="./gradlew"
+      else
+	  GRADLE_EXEC="gradle"
+      fi
+      CLEAN_CMD="${GRADLE_EXEC} clean -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+      BUILD_CMD="${GRADLE_EXEC} clean compileJava -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+  elif [ -f pom.xml ]; then
+      if [ -f mvnw ]; then
+	  chmod +x mvnw
+	  MVN_EXEC="./mvnw"
+      else
+	  MVN_EXEC="mvn"
+      fi
+      # if running on java 8, need /jre at the end of this Maven command
+      if [ "${JAVA_HOME}" = "${JAVA8_HOME}" ]; then
+          CLEAN_CMD="${MVN_EXEC} clean -Djava.home=${JAVA_HOME}/jre"
+          BUILD_CMD="${MVN_EXEC} clean compile -Djava.home=${JAVA_HOME}/jre"
+      else
+          CLEAN_CMD="${MVN_EXEC} clean -Djava.home=${JAVA_HOME}"
+          BUILD_CMD="${MVN_EXEC} clean compile -Djava.home=${JAVA_HOME}"
+      fi
+  else
+      echo "no build file found for ${REPO_NAME}; not calling DLJC"
+      USABLE="no"
+      return
+  fi
+    
+  DLJC_CMD="${DLJC} -t wpi $* -- ${BUILD_CMD}"
+
+  if [ ! "x${TIMEOUT}" = "x" ]; then
+      TMP="${DLJC_CMD}"
+      DLJC_CMD="timeout ${TIMEOUT} ${TMP}"
+  fi
+
+  # Remove old DLJC output.
+  rm -rf dljc-out
+
+  # ensure the project is clean before invoking DLJC
+  eval "${CLEAN_CMD}" < /dev/null
+
+  echo "${DLJC_CMD}"
+
+  # This command also includes "clean"; I'm not sure why it is necessary.
+  eval "${DLJC_CMD}" < /dev/null
+
+  if [[ $? -eq 124 ]]; then
+      echo "dljc timed out for ${DIR}"
+      USABLE="no"
+      return
+  fi
+
+  if [ -f dljc-out/wpi.log ]; then
+      USABLE="yes"
+  else
+      # if this last run was under Java 11, try to run
+      # under Java 8 instead
+      if [ "${JAVA_HOME}" = "${JAVA11_HOME}" ]; then
```

Comment:
This is a funny way to recurse.  I think it would be cleaner to put this logic outside of the `configure_and_exec_dljc` routine; that is, have the main part of the script set JAVA_HOME, then call `configure_and_exect_dljc`, then possibly set JAVA_HOME again and call it a second time.  A minor advantage of this is that it permits you to save the output of both attempts, in case someone wants to investigate them.